### ISSUE TITLE
Traces in monoidal categories and looping in Julia programs

### ISSUE THREAD

**Issue by epatters** [epatters/Catlab.jl#58]
As a followup to #52, we should investigate whether loops in Julia programs can be converted into traces (feedback) in the wiring diagrams. In a cartesian monoidal category, traces are the same as fixed point operators and so are closely related to recursion and looping.

Question first raised on [Twitter](https://twitter.com/laMudri/status/1198003421758021634).

**Comment by epatters** [comment]
Another possibly relevant project is Conal Elliott's [ConCat](https://github.com/conal/concat/), described in the paper [Compiling to categories](http://conal.net/papers/compiling-to-categories/). Tip via @mehalter and @jpfairbanks.

Based on quick glance, I don't see anything about looping/traces (however, this is fairly large project). The circuits example does contain an [`if` construct](https://github.com/conal/concat/blob/86f151eb7555a8738ad3375d6eeb2c01910baf60/examples/src/ConCat/Circuit.hs#L1655). Conditionals are needed for looping to be practical.

The ConCat circuits use the common strategy of representing `if` as a ternary function `if: Bool x A x A -> A`.

**Comment by AriMKatz** [comment]
This package might be useful https://github.com/TuringLang/IRTracker.jl

**Comment by epatters** [comment]
Interesting, thanks!

**Comment by AriMKatz** [comment]
Sure! and it's build on https://github.com/MikeInnes/IRTools.jl, which provides a lower level but presumably more general way of getting the IR for julia code.

**Comment by jpfairbanks** [comment]
Yeah I think there are deep connections between representing programs as wiring diagrams in Catlab and things like AutoDiff. They both turn programs into data structures and then evaluate those programs with different semantics. For AD you want to take a program that computes `f(x)` and map it to a program that computes `f'(x)`. It is interesting to see what level of the Julia compiler process is best for these transformations. Catlab is currently representing programs in monoidal categories with wiring diagrams and doing everything at the AST level with the `@program` macro. I think for things like AD where you are focused on taking existing Julia code and extracting a wiring diagram from it, ie. you are not allowed to tell the users that they have to write in your DSL, the use of IR level code makes a lot of sense.

**Comment by epatters** [comment]
I agree with that, @jpfairbanks. To summarize, I think we need to distinguish two quite different goals:

1. **Julia-esque DSL** to input wiring diagrams. We should keep this as simple as possible, tightly controlling the allowed idioms and avoiding external dependencies.

2. **Arbitrary Julia code** as wiring diagrams. I've prototyped this for [Python](https://github.com/IBM/pyflowgraph) and [R](https://github.com/IBM/rflowgraph), but not for Julia (cf. https://github.com/IBM/semanticflowgraph/issues/11). If we tried something like this, we should investigate IR-level code and packages like Cassette.jl and IRTracker.jl.

**Comment by jpfairbanks** [comment]
Tools for doing the extraction from regular julia code include Cassette and JuliaInterpreter, which can be used to make call trace graphs. https://gist.github.com/mkborregaard/81825c3d370bb4d8dbfe59c3b2ae4b33
